pkg/reconciler/namespace/namespace.go

```diff
@@ -100,55 +103,156 @@ func (c *Controller) reconcileGVR(ctx context.Context, gvr schema.GroupVersionRe
+	klog.Infof("Patching namespace %q status to indicate schedulable=%s", ns.Name, schedulable)
+
+	if schedulable {
+		return c.patchStatus(ctx, ns, func(conditionsSetter conditions.Setter) {
+			conditions.MarkTrue(conditionsSetter, NamespaceScheduled)
+		})
 	}
+	return c.patchStatus(ctx, ns, func(conditionsSetter conditions.Setter) {
+		conditions.MarkFalse(conditionsSetter, NamespaceScheduled, NamespaceReasonUnschedulable,
+			conditionsv1alpha1.ConditionSeverityNone, // NamespaceCondition doesn't support severity
+			"No clusters are available to schedule Namespaces to.")
+	})
+}
 
-	clusterName := ns.Labels[clusterLabel]
-	if clusterName == "" {
-		// Namespace is not assigned to a cluster; assign one.
-		// First, list all clusters.
-		cls, err := c.clusterLister.List(labels.Everything())
-		if err != nil {
-			return err
-		}
+// assignCluster attempts to schedule a namespace to a random cluster. If no cluster is
+// not available, any existing cluster assignment will be cleared.
+func (c *Controller) assignCluster(ctx context.Context, ns *corev1.Namespace) error {
+	clusters, err := c.clusterLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
 
-		// TODO: Filter out un-Ready clusters so a namespace doesn't
-		// get assigned to an un-Ready cluster.
+	oldClusterName := ns.Labels[clusterLabel]
 
-		if len(cls) == 0 {
-			// TODO set status, on namespace and all resources in namespace.
-			return fmt.Errorf("no clusters")
-		}
+	// TODO: Filter out un-Ready clusters so a namespace doesn't
+	// get assigned to an un-Ready cluster.
 
+	newClusterName := ""
+	if len(clusters) > 0 {
 		// Select a cluster at random.
-		cl := cls[rand.Intn(len(cls))]
-		old, new := "", cl.Name
+		cluster := clusters[rand.Intn(len(clusters))]
+		newClusterName = cluster.Name
+	}
 
-		// Patch the namespace.
-		klog.Infof("Patching to update cluster assignment for namespace %s: %q -> %q", ns.Name, old, new)
-		if _, err := c.namespaceClient.Patch(ctx, ns.Name, types.MergePatchType, clusterLabelPatchBytes(new), metav1.PatchOptions{}); err != nil {
+	if oldClusterName != newClusterName {
+		klog.Infof("Patching to update cluster assignment for namespace %s: %q -> %q", ns.Name, oldClusterName, newClusterName)
+		if _, err := c.namespaceClient.Patch(ctx, ns.Name, types.MergePatchType, clusterLabelPatchBytes(newClusterName), metav1.PatchOptions{}); err != nil {
 			return err
 		}
+
 	}
 
-	// Check if the cluster reports as Ready.
-	cl, err := c.clusterLister.Get(clusters.ToClusterAwareKey(lclusterName, clusterName))
+	return nil
+}
+
+// isValidCluster checks whether the given cluster name exists and is valid for the
+// purposes of scheduling a namespace to it.
+func (c *Controller) isValidCluster(ctx context.Context, lclusterName, clusterName string) (bool, error) {
+	cluster, err := c.clusterLister.Get(clusters.ToClusterAwareKey(lclusterName, clusterName))
+	if apierrors.IsNotFound(err) {
+		klog.Infof("Cluster %q does not exist", clusterName)
```

Comment:
this seems to be noisy. I assume this func can be called pretty often.

Reply:
Any given cluster is likely to be assigned to multiple namespaces, so that if a cluster is deleted or goes unready many namespaces will be affected and hit these log lines. Isn't that desirable though? The log could show 'cluster foo does not exist' and then 'scheduling namespace foo to other cluster'.